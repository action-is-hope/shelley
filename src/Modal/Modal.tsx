"use client";
import type { TransitionProps } from "react-transition-group/Transition";
import type { ReactFocusOnProps } from "react-focus-on/dist/es5/types";
import React, { useRef, forwardRef } from "react";
import { createPortal } from "react-dom";
import { CSSTransition } from "react-transition-group";
import { FocusOn } from "react-focus-on";
import { st, classes } from "./modal.st.css";
import { mergeProps } from "react-aria";

export function composeEventHandlers(
  theirHandler: React.MouseEventHandler<HTMLDivElement>,
  ourHandler: React.MouseEventHandler<HTMLDivElement>
) {
  return (event: React.MouseEvent<HTMLDivElement>) => {
    theirHandler && theirHandler(event);
    if (!event.defaultPrevented) {
      return ourHandler(event);
    }
  };
}

export type TransitionType = "up" | "zoom" | string | false;

export interface ModalProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Controls whether the dialog is open or not. */
  isOpen?: boolean;
  /** Accepts any renderable content. */
  children?: React.ReactNode;
  /** Add a class to the content div. */
  contentClassName?: string;
  /** Disables the backdrop click dismiss */
  disableBackdropClick?: boolean;
  /** Disables the EscapeKey dismiss */
  disableEscapeKey?: boolean;
  /** Disables the FocusLock - shortcut to focusOnProps */
  disableFocusLock?: boolean;
  /** Props for the internal `FocusOn` component see - https://github.com/theKashey/react-focus-on#api */
  focusOnProps?: Pick<
    ReactFocusOnProps,
    Exclude<keyof ReactFocusOnProps, "children">
  >;
  /**
   * Set the initial focused element via a ref. By default the
   * 'first focusable element' will receive focus when the dialog
   * opens.
   */
  initialFocusRef?:
    | React.RefObject<HTMLButtonElement>
    | React.RefObject<HTMLInputElement>;
  /** Fires when the backdrop is clicked. */
  onBackdropClick?: () => void;
  /**
   * This function is called whenever the user hits "Escape" or clicks outside
   * the dialog. _It's important to close the dialog `onDismiss`_.
   *
   * The only time you shouldn't close the dialog on dismiss is when the dialog
   * requires a choice and none of them are "cancel". For example, perhaps two
   * records need to be merged and the user needs to pick the surviving record.
   * Neither choice is less destructive than the other, so in these cases you
   * may want to alert the user they need to a make a choice on dismiss instead
   * of closing the dialog.
   */
  onDismiss?: () => void;
  /**
   * The selector of the element that the menu should render inside of.
   * @default 'body'
   */
  portalSelector?: string | false;
  /**
   * Type of transition
   * @default "up"
   */
  transition?: TransitionType;
  /** Props for the internal `CSSTransition` component see - https://reactcommunity.org/react-transition-group/css-transition */
  transitionProps?: Omit<TransitionProps, "addEndListener">;
  variant?: "fixed" | string | false;
  disableModalBackdropBlur?: boolean;
  /** Add predefined data-id to ease testing or analytics. */
  "data-id"?: string;
}

function Modal(props: ModalProps, ref?: React.Ref<HTMLDivElement>) {
  const {
    className: classNameProp,
    children,
    contentClassName,
    portalSelector = "body",
    focusOnProps: focusOnPropsInput,
    initialFocusRef,
    isOpen,
    onBackdropClick = () => null,
    onDismiss = () => null,
    onMouseDown = () => null,
    transitionProps: transitionPropsFromProps,
    variant = "fixed",
    transition = "zoom",
    disableFocusLock,
    disableEscapeKey,
    disableBackdropClick,
    disableModalBackdropBlur,
    "data-id": dataId,
    ...rest
  } = props;
  const mouseDownTarget = useRef<EventTarget | null>(null);

  const activateFocusLock = React.useCallback(() => {
    if (initialFocusRef && initialFocusRef.current) {
      initialFocusRef.current.focus();
    }
  }, [initialFocusRef]);

  const handleBackdropClick = (event: React.MouseEvent<HTMLDivElement>) => {
    !disableBackdropClick && onDismiss();
    if (mouseDownTarget.current === event.target) {
      event.stopPropagation();
      !disableBackdropClick && onDismiss();
    }
  };

  const handleMouseDown = (event: React.MouseEvent) => {
    mouseDownTarget.current = event.target;
  };

  const transitionPropsDefault = {
    unmountOnExit: true,
    ...transitionPropsFromProps,
    onEntering: () => {
      /* eslint-disable  @typescript-eslint/no-unsafe-call */
      typeof transitionPropsFromProps?.onEntering === "function" &&
        transitionPropsFromProps?.onEntering();
      !disableModalBackdropBlur &&
        document.body.classList.add(classes.blurBackground);
    },
    onExiting: () => {
      /* eslint-disable  @typescript-eslint/no-unsafe-call */
      typeof transitionPropsFromProps?.onExiting === "function" &&
        transitionPropsFromProps?.onExiting();
      !disableModalBackdropBlur &&
        document.body.classList.remove(classes.blurBackground);
    },
  };

  const transitionProps = transitionPropsFromProps
    ? mergeProps(transitionPropsDefault, transitionPropsFromProps)
    : transitionPropsDefault;

  const focusOnProps = {
    focusLock: !disableFocusLock,
    /**
     * If the dialog is persistent in the DOM via `unmountOnExit: false` then
     * disable the isolation lock else everything else will have aria-hidden="true".
     * noIsolation cannot be dynamically toggled with isOpen.
     * Possible work around to use shards?
     */
    noIsolation: !transitionProps.unmountOnExit,
    ...focusOnPropsInput,
    onEscapeKey: (event: Event) => {
      !disableEscapeKey && onDismiss();
      focusOnPropsInput?.onEscapeKey && focusOnPropsInput?.onEscapeKey(event);
    },
    onActivation: activateFocusLock,
  };

  const nodeRef = useRef(null);
  const modal = (
    <CSSTransition
      in={isOpen}
      timeout={190}
      {...transitionProps}
      nodeRef={nodeRef}
    >
      <>
        <div
          className={st(
            classes.root,
            {
              variant: variant || undefined,
              transition: transition || undefined,
            },
            classNameProp
          )}
          ref={nodeRef}
          data-id={dataId}
        >
          <FocusOn {...focusOnProps}>
            <div
              aria-hidden="true"
              className={classes.backdrop}
              data-id={dataId ? `${dataId}-backdrop` : undefined}
              onClick={composeEventHandlers(
                onBackdropClick,
                handleBackdropClick
              )}
            ></div>
            <div
              data-id={dataId ? `${dataId}-content` : undefined}
              className={st(classes.content, contentClassName)}
              onMouseDown={composeEventHandlers(onMouseDown, handleMouseDown)}
              ref={ref}
              {...rest}
            >
              {children}
            </div>
          </FocusOn>
        </div>
      </>
    </CSSTransition>
  );
  // If portalSelector then render inside portal elso render inline.
  // @todo check element exists.
  return portalSelector
    ? createPortal(modal, document.querySelector(portalSelector) as HTMLElement)
    : modal;
}

/**
 * Modal provides the structure for a Modal including transitions and
 * focus locks required to keep a user focused inside active dialogs without
 * being able to 'wonder about' behind the backdrop.
 */

const _Modal = forwardRef(Modal);
export { _Modal as Modal };
