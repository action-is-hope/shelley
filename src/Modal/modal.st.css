@st-namespace "Modal";

.root {
  -st-states:
    variant(string),
    transition(string);
}

.root {
  display: flex;
  align-items: center;
  justify-content: center;
  /*  */
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 2000;
  --transiton-slide-from-start: translateX(-100%);
  --transiton-slide-from-end: translateX(100%);
}

[dir=rtl] .root {
  --transiton-slide-from-start: translateX(100%);
  --transiton-slide-from-end: translateX(-100%);
}

.root:variant(fixed) {
  position: fixed;
}

.backdrop {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

.content {
  position: relative;
}

.blurBackground [data-focus-on-hidden]:not(.root, .root *) {
  opacity: .5;
  filter: blur(2px);
}

/* = Content Transitions */

/* Map transition lifecycle classes to global CSSTransitionGroup classes. */
.enter {
  -st-global: '.enter';
}

.enterActive {
  -st-global: '.enter-active';
}

.enterDone {
  -st-global: '.enter-done';
}

.exit {
  -st-global: '.exit';
}

.exitActive {
  -st-global: '.exit-active';
}

.exitDone {
  -st-global: '.exit-done';
  display: none;
}

/* Background transition. */
.root.enter .backdrop {
  opacity: 0;
}

.root.enterActive .backdrop {
  opacity: 1;
  transition: opacity 190ms;
}

.root.enterDone .backdrop {
  opacity: 1;
}

.root.exit .backdrop {
  opacity: 1;
}

.root.exitActive .backdrop {
  opacity: 0;
  transition: opacity 300ms;
}

/* = Zoom */
.root:transition(zoom).enter .content {
  transform: scale(0);
  opacity: 0;
}

.root:transition(zoom).enterActive .content {
  transform: scale(1);
  opacity: 1;
  transition: transform 190ms, opacity 190ms;
}

.root:transition(zoom).enterDone .content {
  transform: scale(1);
  opacity: 1;
}

.root:transition(zoom).exit .content {
  transform: scale(1);
  opacity: 1;
}

.root:transition(zoom).exitActive .content {
  transform: scale(0);
  opacity: 0;
  transition: transform 190ms, opacity 190ms;
}

/* = slideUp */
.root:transition(slideUp).enter .content {
  transform: translateY(calc(100vh / 2 + 100%));
  opacity: 0;
}

.root:transition(slideUp).enterActive .content {
  transform: translateY(0);
  opacity: 1;
  transition: transform 190ms, opacity 190ms;
}

.root:transition(slideUp).enterDone .content {
  transform: translateY(0);
  opacity: 1;
}

.root:transition(slideUp).exit .content {
  transform: translateY(0);
  opacity: 1;
}

.root:transition(slideUp).exitActive .content {
  transform: translateY(calc(100vh / 2 + 100%));
  opacity: 0;
  transition: transform 190ms, opacity 190ms;
}

/* = slideFromStart = */
.root:transition(slideFromStart).enter .content {
  transform: var(--transiton-slide-from-start);
  opacity: 0;
}

.root:transition(slideFromStart).enterActive .content {
  transform: translateX(0);
  opacity: 1;
  transition: transform 190ms, opacity 190ms;
}

.root:transition(slideFromStart).enterDone .content {
  transform: translateX(0);
  opacity: 1;
}

.root:transition(slideFromStart).exit .content {
  transform: translateX(0);
  opacity: 1;
}

.root:transition(slideFromStart).exitActive .content {
  transform: var(--transiton-slide-from-start);
  opacity: 0;
  transition: transform 190ms, opacity 190ms;
}

/* = slideFromEnd = */
.root:transition(slideFromEnd).enter .content {
  transform: var(--transiton-slide-from-end);
  opacity: 0;
}

.root:transition(slideFromEnd).enterActive .content {
  transform: translateX(0);
  opacity: 1;
  transition: transform 190ms, opacity 190ms;
}

.root:transition(slideFromEnd).enterDone .content {
  transform: translateX(0);
  opacity: 1;
}

.root:transition(slideFromEnd).exit .content {
  transform: translateX(0);
  opacity: 1;
}

.root:transition(slideFromEnd).exitActive .content {
  transform: var(--transiton-slide-from-end);
  opacity: 0;
  transition: transform 190ms, opacity 190ms;
}
