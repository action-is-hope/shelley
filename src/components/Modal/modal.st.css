/* modal.st.css */
@namespace "Modal";

.root {
  -st-states:
    variant(string),
    transition(string);
}

.root {
  display: flex;
  align-items: center;
  justify-content: center;
  /*  */
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 2000;
}

.root:variant(fixed) {
  position: fixed;
}

.backdrop {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

.content {}

/* .transition {} */

.blurBackground [data-focus-on-hidden] {
  opacity: .5;
  filter: blur(2px);
}

/* = Content Transitions */

/* Map transition liftcycle classes to global CSSTransistionGroup classes. */
.enter {
  -st-global: '.enter'
}

.enterActive {
  -st-global: '.enter-active'
}

.enterDone {
  -st-global: '.enter-done'
}

.exit {
  -st-global: '.exit'
}

.exitActive {
  -st-global: '.exit-active'
}

.exitDone {
  -st-global: '.exit-done';
  display: none;
}

/* Background transition. */
.root.enter .backdrop {
  opacity: 0;
}

.root.enterActive .backdrop {
  opacity: 1;
  transition: opacity 190ms;
}

.root.enterDone .backdrop {
  opacity: 1;
}

.root.exit .backdrop {
  opacity: 1;
}

.root.exitActive .backdrop {
  opacity: 0;
  transition: opacity 300ms;
}

/* = up */
.root:transition(up).enter .content {
  transform: translateY(calc(100vh / 2 + 100%));
  opacity: 0;
}

.root:transition(up).enterActive .content {
  transform: translateY(0);
  opacity: 1;
  transition: transform 190ms, opacity 190ms;
}

.root:transition(up).enterDone .content {
  transform: translateY(0);
  opacity: 1;
}

.root:transition(up).exit .content {
  transform: translateY(0);
  opacity: 1;
}

.root:transition(up).exitActive .content {
  transform: translateY(calc(100vh / 2 + 100%));
  opacity: 0;
  transition: transform 190ms, opacity 190ms;
}

/* = Zoom */

.root:transition(zoom).enter .content {
  transform: scale(0);
  opacity: 0;
}

.root:transition(zoom).enterActive .content {
  transform: scale(1);
  opacity: 1;
  transition: transform 190ms, opacity 190ms;
}

.root:transition(zoom).enterDone .content {
  transform: scale(1);
  opacity: 1;
}

.root:transition(zoom).exit .content {
  transform: scale(1);
  opacity: 1;
}

.root:transition(zoom).exitActive .content {
  transform: scale(0);
  opacity: 0;
  transition: transform 190ms, opacity 190ms;
}
