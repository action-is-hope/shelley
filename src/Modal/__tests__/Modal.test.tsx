import { Modal } from "../Modal";
import renderer from "react-test-renderer";

describe("Modal", () => {
  it("renders correctly with classname", () => {
    const tree = renderer
      .create(
        <Modal
          isOpen
          transition={false}
          data-id="modal-test"
          className={"modal-classname-test"}
          portalSelector={false}
        >
          <div>Content</div>
        </Modal>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
