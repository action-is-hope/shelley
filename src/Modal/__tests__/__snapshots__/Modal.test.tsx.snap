// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly with classname 1`] = `
<div
  className="Modal__Test__root Modal__Test---variant-5-fixed modal-classname-test"
  data-id="modal-test"
>
  <div
    data-focus-guard={true}
    style={
      {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />
  <div
    data-focus-lock-disabled="disabled"
    onBlur={[Function]}
    onFocus={[Function]}
    onScrollCapture={[Function]}
    onTouchMoveCapture={[Function]}
    onWheelCapture={[Function]}
  >
    <div
      aria-hidden="true"
      className="Modal__Test__backdrop"
      data-id="modal-test-backdrop"
      onClick={[Function]}
    />
    <div
      className="Modal__Test__content"
      data-id="modal-test-content"
      onMouseDown={[Function]}
    >
      <div>
        Content
      </div>
    </div>
  </div>
  <div
    data-focus-guard={true}
    style={
      {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />
</div>
`;
